src/art.rs

```diff
@@ -0,0 +1,1827 @@
+            NodeType::Node256(n) => NodeType::Node256(n.clone()),
+        }
+    }
+}
+
+// Default implementation for the Tree struct
+impl<P: KeyTrait, V: Clone> Default for Tree<P, V> {
+    fn default() -> Self {
+        Tree::new()
+    }
+}
+
+impl<P: KeyTrait + Clone, V: Clone> Node<P, V> {
+    /// Creates a new Twig node with a given prefix, key, value, and timestamp.
+    ///
+    /// Constructs a new Twig node using the provided prefix, key, and value. The timestamp
+    /// indicates the time of the insertion.
+    ///
+    /// # Parameters
+    ///
+    /// - `prefix`: The common prefix for the node.
+    /// - `key`: The key associated with the Twig node.
+    /// - `value`: The value to be associated with the key.
+    /// - `ts`: The timestamp when the value was inserted.
+    ///
+    /// # Returns
+    ///
+    /// Returns a new `Node` instance with a Twig node containing the provided key, value, and timestamp.
+    ///
+    #[inline]
+    pub(crate) fn new_twig(prefix: P, key: P, value: V, ts: u64) -> Node<P, V> {
+        // Create a new TwigNode instance using the provided prefix and key.
+        let mut twig = TwigNode::new(prefix, key);
+
+        // Insert the provided value into the TwigNode along with the timestamp.
+        twig.insert_mut(value, ts);
+
+        // Return a new Node instance encapsulating the constructed Twig node.
+        Self {
+            node_type: NodeType::Twig(twig),
+        }
+    }
+
+    /// Creates a new inner Node4 node with the provided prefix.
+    ///
+    /// Constructs a new Node4 node using the provided prefix. Node4 is an inner node
+    /// type that can store up to 4 child pointers and uses arrays for keys and pointers.
+    ///
+    /// # Parameters
+    ///
+    /// - `prefix`: The common prefix for the Node4 node.
+    ///
+    /// # Returns
+    ///
+    /// Returns a new `Node` instance with an empty Node4 node.
+    ///
+    #[inline]
+    #[allow(dead_code)]
+    pub(crate) fn new_node4(prefix: P) -> Self {
+        // Create a new FlatNode instance using the provided prefix.
+        let flat_node = FlatNode::new(prefix);
+
+        // Create a new Node4 instance with the constructed FlatNode.
+        Self {
+            node_type: NodeType::Node4(flat_node),
+        }
+    }
+
+    /// Checks if the current node is full based on its type.
+    ///
+    /// Determines if the current node is full by comparing the number of children to its
+    /// capacity, which varies based on the node type.
+    ///
+    /// # Returns
+    ///
+    /// Returns `true` if the node is full, `false` otherwise.
+    ///
+    #[inline]
+    fn is_full(&self) -> bool {
+        match &self.node_type {
+            NodeType::Node4(n) => self.num_children() >= n.size(),
+            NodeType::Node16(n) => self.num_children() >= n.size(),
+            NodeType::Node48(n) => self.num_children() >= n.size(),
+            NodeType::Node256(_) => self.num_children() > 256,
```

Comment:
I'm curious why `>` is used instead of `>=` and why this doesn't use `n.size()` (which seems to return 256) like the other cases.

Reply:
Hi @finnbear, thank you for taking time to review this. Yes you are right, it should be `>=`

I was thinking of using constants for 256 and 48 rather than call size to avoid unnecessary call of a function as both have fixed sizes unlike node4/16 which using the same underlying flatnode structure

Reply:
> to avoid unnecessary call of a function

Just use `#[inline(always)]` on `size` if you're worried about that overhead (not that you should be. since it's within the same crate, Rust will almost certainly optimize it regardless).

Reply:
I was indeed thinking about the function call overhead, but there you have it, something new that you taught me! Thanks for the suggestion(s)!